cardinal-components-base/src/main/java/dev/onyxstudios/cca/internal/base/ComponentRegistryImpl.java

```diff
@@ -1,162 +1,164 @@
+import net.minecraft.util.Identifier;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nullable;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public final class ComponentRegistryImpl implements ComponentRegistry {
+
+    // used by generated classes
+    public static ComponentType<?> byRawId(@Nonnegative int rawId) {
+        ComponentRegistryImpl registry = (ComponentRegistryImpl) INSTANCE;
+        ComponentType<?> ret = registry.get(rawId);
+        if (ret == null) {
+            for (Object2IntMap.Entry<Identifier> entry : registry.id2Raw.object2IntEntrySet()) {
+                if (entry.getIntValue() == rawId) {
+                    throw new IllegalStateException("The component type for '" + entry.getKey() + "'  was not registered");
+                }
+            }
+            throw new IllegalArgumentException("Invalid raw id " + rawId);
+        }
+        return ret;
+    }
+
+    private final ComponentTypeAccess access;
+    private final Object2IntMap<Identifier> id2Raw;
+    private ComponentType<?>[] raw2Types;
+    private int nextRawId = 0;
+    private int size;
+
+    public ComponentRegistryImpl(ComponentTypeAccess access) {
+        this.access = access;
+        this.id2Raw = new Object2IntOpenHashMap<>(16);
+        this.raw2Types = new ComponentType[16];
+        this.id2Raw.defaultReturnValue(-1);
+    }
+
+    @Override
+    public synchronized <T extends Component> ComponentType<T> registerIfAbsent(Identifier componentId, Class<T> componentClass) {
+        Preconditions.checkArgument(Component.class.isAssignableFrom(componentClass), "Component interface must extend " + Component.class.getCanonicalName());
+        // make sure 2+ components cannot get registered at the same time
+        int rawId = this.assignRawId(componentId);
+        @SuppressWarnings("unchecked")
+        ComponentType<T> existing = (ComponentType<T>) this.get(rawId);
+        if (existing != null) {
+            if (existing.getComponentClass() != componentClass) {
+                throw new IllegalStateException("Registered component " + componentId + " twice with 2 different classes: " + existing.getComponentClass() + ", " + componentClass);
+            }
+            return existing;
+        } else {
+            ComponentType<T> registered;
+            Class<? extends ComponentType<?>> generated = CcaBootstrap.INSTANCE.getGeneratedComponentTypeClass(componentId);
+            if (generated != null) {
+                registered = this.instantiateStaticType(generated, componentId, componentClass, rawId);
+            } else {
+                registered = this.access.create(componentId, componentClass, rawId);
+            }
+            if (this.raw2Types.length <= rawId) {
+                ComponentType<?>[] newArray = new ComponentType[this.raw2Types.length + 16];

```

Comment:
`this.raw2Types.length + 16` is actually still invalid, because static components can allocate raw ids that do not get claimed in order.